File: 152. Maximum Product Subarray.py
class Solution(object):
    def maxProduct(self, nums):
        """
        :type nums: List[int]
        :rtype: int
        """
        T = [[0 for j in xrange(2)] for i in xrange(len(nums) + 1)]
        T[1][0], T[1][1], res = nums[0], nums[0], nums[0]
        for i in xrange(2, len(nums) + 1):
            T[i][0] = max(T[i-1][0] * nums[i-1], T[i-1][1] * nums[i-1], nums[i-1])
            T[i][1] = min(T[i-1][0] * nums[i-1], T[i-1][1] * nums[i-1], nums[i-1])
            res = max(T[i][0], res)
        return res

File: 152. Maximum Product Subarray.cpp
class Solution {
public:
    int maxProduct(vector<int>& nums) {
        vector <vector <int>> T (nums.size() + 1, vector<int>(2, 0));
        T[1][0] = nums[0];
        T[1][1] = nums[0];
        int res = nums[0];
        for (int i = 2; i <= nums.size(); i++)
        {
            int temp = T[i-1][0] * nums[i-1] > T[i-1][1] * nums[i-1] ? T[i-1][0] * nums[i-1] : T[i-1][1] * nums[i-1];
            T[i][0] = temp > nums[i-1] ? temp : nums[i-1];
            temp = T[i-1][0] * nums[i-1] < T[i-1][1] * nums[i-1] ? T[i-1][0] * nums[i-1] : T[i-1][1] * nums[i-1];
            T[i][1] = temp < nums[i-1] ? temp : nums[i-1];
            res = res > T[i][0] ? res : T[i][0];
        }
        return res;
    }
};
